pkg/sink/pulsar/factory.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulsar
+
+import (
+	"fmt"
+
+	"github.com/apache/pulsar-client-go/pulsar"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/model"
+	"go.uber.org/zap"
+)
+
+// FactoryCreator defines the type of factory creator.
+type FactoryCreator func(config *Config, changefeedID model.ChangeFeedID) (pulsar.Client, error)
+
+// NewCreatorFactory returns a factory implemented based on kafka-go
+func NewCreatorFactory(config *Config, changefeedID model.ChangeFeedID) (pulsar.Client, error) {
+	co := pulsar.ClientOptions{
+		URL: config.URL,
+		// MetricsRegisterer: pulsarMetric.GetMetricRegistry(),
+		CustomMetricsLabels: map[string]string{
+			"changefeed": changefeedID.ID,
+			"namespace":  changefeedID.Namespace,
+		},
+		ConnectionTimeout: config.ConnectionTimeout,
+		OperationTimeout:  config.OperationTimeout,
+	}
+	var err error
+
+	co.Authentication, err = setupAuthentication(config)
+	if err != nil {
+		log.L().Error("setup pulsar authentication fail", zap.Error(err))
+		return nil, err
+	}
+
+	pulsarClient, err := pulsar.NewClient(co)
+	if err != nil {
+		log.L().Error("cannot connect to pulsar", zap.Error(err))
+		return nil, err
+	}
+	return pulsarClient, nil
+}
+
+func setupAuthentication(config *Config) (pulsar.Authentication, error) {
+	if len(config.AuthenticationToken) > 0 {
+		return pulsar.NewAuthenticationToken(config.AuthenticationToken), nil
+	} else if len(config.TokenFromFile) > 0 {
+		return pulsar.NewAuthenticationTokenFromFile(config.TokenFromFile), nil
+	} else if len(config.BasicUserName) > 0 && len(config.BasicPassword) > 0 {
+		return pulsar.NewAuthenticationBasic(config.BasicUserName, config.BasicPassword)
+	} else if len(config.OAuth2) >= 5 {
+		return pulsar.NewAuthenticationOAuth2(config.OAuth2), nil
+	} else if len(config.TLSCertificatePath) > 0 && len(config.TLSPrivateKeyPath) > 0 {
```

Comment:
It seems that the TLS encryption config can live with authentication config at the same time. Am I right?

Reply:
OAuth2 support encryption